markdown_it/cli/parse.py

```diff
@@ -15,7 +17,7 @@
 version_str = "markdown-it-py [version {}]".format(__version__)
 
 
-def main(args: Optional[Sequence[str]] = None) -> int:
+def main(args: Sequence[str] | None = None) -> int:
```

Comment:
dammit, made this comment ages ago, but didn't realise it was pending

Reply:
This works perfectly!

It would be a syntax error but `from __future__ import annotations` makes it not error.

And `mypy` doesn't care what our minimum supported Python version is.

Reply:
Not sure how mypy does that 🤷 (are you sure it actually picks them up, and not silently fails)

On python 3.10 you can do:

```python
In [1]: def bar() -> int | str:
   ...:     pass
In [2]: from typing import get_type_hints
In [3]: get_type_hints(bar)
Out[3]: {'return': int | str}
```

but on python 3.8, you get:

```python
In [1]: from __future__ import annotations
In [2]: from typing import get_type_hints
In [3]: def bar() -> int | str:
   ...:     pass
   ...: 
In [4]: get_type_hints(bar)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-348bd5169782> in <module>
----> 1 get_type_hints(bar)

~/Documents/GitHub/aiida_core_develop/.tox/py38-pre-commit/lib/python3.8/typing.py in get_type_hints(obj, globalns, localns)
   1262         if isinstance(value, str):
   1263             value = ForwardRef(value)
-> 1264         value = _eval_type(value, globalns, localns)
   1265         if name in defaults and defaults[name] is None:
   1266             value = Optional[value]

~/Documents/GitHub/aiida_core_develop/.tox/py38-pre-commit/lib/python3.8/typing.py in _eval_type(t, globalns, localns)
    268     """
    269     if isinstance(t, ForwardRef):
--> 270         return t._evaluate(globalns, localns)
    271     if isinstance(t, _GenericAlias):
    272         ev_args = tuple(_eval_type(a, globalns, localns) for a in t.__args__)

~/Documents/GitHub/aiida_core_develop/.tox/py38-pre-commit/lib/python3.8/typing.py in _evaluate(self, globalns, localns)
    516                 localns = globalns
    517             self.__forward_value__ = _type_check(
--> 518                 eval(self.__forward_code__, globalns, localns),
    519                 "Forward references must evaluate to types.",
    520                 is_argument=self.__forward_is_argument__)

<string> in <module>

TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

Reply:
I'm sure it works. Mypy is a static type checker. It doesn't run the code.

Reply:
hmm, this doesn't work for `def bar() -> list[str]: pass` either, and also `typing_extensions.get_type_hints` didn't work, I guess mypy has some other way

Reply:
yep: https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-generic-builtins/https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-x-y-syntax-for-unions
I doubt that anyone will need to dynamically evaluate these types, but something to bear in mind

Reply:
> It doesn't run the code it type checks

Not sure about sphinx autodoc though

Reply:
I thought sphinx-autodoc doesn't do anything with type annotations unless [sphinx-autodoc-typehints](https://github.com/tox-dev/sphinx-autodoc-typehints) is in use (it isn't)?

And the plugin certainly supports what we do in this PR: https://github.com/tox-dev/sphinx-autodoc-typehints/pull/184

Reply:
Yeh exactly, looks all good 👍; I just wanted to make sure I understood the implications using future annotations, before rolling them out everywhere